src/app.rs

```diff
@@ -1,133 +1,496 @@
+            Box::new(Bank {
+                store: SubStore::new(store.clone(), prefix::Bank),
+            }),
+            Box::new(Ibc {
+                store: SubStore::new(store.clone(), prefix::Ibc),
+                client_counter: 0,
+            }),
+        ];
         Self {
-            state: Arc::new(RwLock::new(BaseCoinState::default())),
+            store,
+            modules: Arc::new(RwLock::new(modules)),
         }
     }
+
+    pub(crate) fn sub_store<I: Identifiable>(
+        &self,
+        prefix: I,
+    ) -> SubStore<SharedStore<WalStore<S>>, I> {
+        SubStore::new(self.store.clone(), prefix)
+    }
 }
 
-impl Application for BaseCoinApp {
-    fn info(&self, request: RequestInfo) -> ResponseInfo {
-        debug!(
-            "Got info request. Tendermint version: {}; Block version: {}; P2P version: {}",
-            request.version, request.block_version, request.p2p_version
-        );
+impl<S> BaseCoinApp<S> {
+    // try to deliver the message to all registered modules
+    // if `module.deliver()` returns `Error::not_handled()`, try next module
+    // Return:
+    // * other errors immediately OR
+    // * `Error::not_handled()` if all modules return `Error::not_handled()`
+    // * events from first successful deliver call OR
+    fn deliver_msg(&self, message: Msg) -> Result<Vec<Event>, Error> {
+        let mut modules = self.modules.write().unwrap();
+        let mut handled = false;
+        let mut events = vec![];
+
+        for m in modules.iter_mut() {
+            match m.deliver(message.clone().into()) {
+                Ok(mut msg_events) => {
+                    events.append(&mut msg_events);
+                    handled = true;
+                    break;
+                }
+                Err(Error(ErrorDetail::NotHandled(_), _)) => continue,
+                Err(e) => return Err(e),
+            };
+        }
+
+        if handled {
+            Ok(events)
+        } else {
+            Err(Error::not_handled())
+        }
+    }
+}
 
+impl<S: ProvableStore + 'static> Application for BaseCoinApp<S> {
+    fn info(&self, request: RequestInfo) -> ResponseInfo {
         let (last_block_height, last_block_app_hash) = {
-            let state = self.state.read().unwrap();
-            (state.height(), state.hash())
+            let state = self.store.read().unwrap();
+            (state.current_height() as i64, state.root_hash())
         };
-
+        debug!(
+            "Got info request. Tendermint version: {}; Block version: {}; P2P version: {}, {:?}, {:?}",
+            request.version, request.block_version, request.p2p_version, last_block_height, last_block_app_hash
+        );
         ResponseInfo {
             data: "basecoin-rs".to_string(),
             version: "0.1.0".to_string(),
             app_version: 1,
             last_block_height,
-            last_block_app_hash,
+            last_block_app_hash: vec![],
```

Comment:
Any particular reason why this isn't set to the local `last_block_app_hash` variable value?

Reply:
Nice catch! Done! 👍